Multiline: Spaces to Newlines

radian628 opened the pull request:
Adds a setting to the Multiline plugin that converts all instances of three consecutive spaces `   ` to newlines. Also adds a setting to disable the default automatic multilining behavior. In other words, stuff like this is now possible:
![image](https://github.com/DesModder/DesModder/assets/70352880/fa05ee2b-a16d-43be-a017-2fa8ceef57bc)

jared-hughes commented:
- Missing localization for `multiline-opt-determineLineBreaksAutomatically-name`
- If `determineLineBreaksAutomatically` is false, then should `automaticallyMultilinify` be false? Or is it needed to process the spaces?
- Shift-enter is also used by hide-errors to hide the errors of the current expression. Looks like it was never documented, so I suspect low usage. Might want to entirely remove the code to hide errors on shift-enter. (We need a keyboard shortcut manager)
- It's reasonable to have `spacesToNewlines` enabled by default. Triple-spaces are rare enough.

radian628 commented:
> Missing localization for `multiline-opt-determineLineBreaksAutomatically-name`

Nice catch!

> If `determineLineBreaksAutomatically` is false, then should `automaticallyMultilinify` be false? Or is it needed to process the spaces?

If the former is false, the latter should really be false too. With the Shift+Enter approach, I'm operating under the assumption that any time someone does Shift+Enter, it implies that they want the expressions multilined. Also, are these names/descriptions clear enough? I feel like they have potential to be confused.

> Shift-enter is also used by hide-errors to hide the errors of the current expression. Looks like it was never documented, so I suspect low usage. Might want to entirely remove the code to hide errors on shift-enter. (We need a keyboard shortcut manager)

For sure lol yeah. I didn't know about the hide-errors Shift+enter either. If I have time I'd be happy to make a keyboard shortcut manager. We could add our own section of configurable hotkeys to the current `Ctrl+/` page.

> It's reasonable to have `spacesToNewlines` enabled by default. Triple-spaces are rare enough.

My only concern with this is that I don't want hand-aligned graphs to be mangled by the automatic multilining if someone who makes a hand-aligned graph shares it with someone who has automatic multilining enabled. To get around this we could use metadata: There could be a boolean flag in the graphwide metadata that states something to the effect of "spacesToNewlines is used at least once on this page, and automatic multilining is disabled!" This would prompt a toast to be generated when a user visits the page *iff* they have automatic multilining enabled. The toast would say something like "This graph was made with the Multiline Expressions Plugin's Spaces to Newlines feature. Consider disabling 'Determine Line Breaks Automatically' to prevent extra line breaks from altering the formatting." Alternatively, we could have _another_ setting that automatically toggles/overrides Multiline settings like this depending on the graph.

radian628 commented:
Alright I've implemented it: There's a new setting (enabled by default) that will remove any automatically-generated line breaks if there's any manual line breaks (triple spaces) in an expression. This should hopefully prevent hand-aligned expressions from getting messed up if someone with automatic aligning enabled checks it out.

jared-hughes commented:
Thoughts on the following renames?
- "Preserve Shift+Enter'd Exprs" → "Skip expressions with triple spaces"
- "Auto Determine Linebreaks" → "Auto Insert Linebreaks"
- "Automatically Multilinify" → "Insert Linebreaks While Typing"

jared-hughes commented on src/plugins/multiline/index.ts:
```diff
@@ -18,6 +18,14 @@ function focusmq(mq: MathQuillField | undefined) {
 
+function isNextToTripleSpaceLineBreak(mq: MathQuillField, dir: 1 | -1) {
+  return (
+    getController(mq).cursor[dir]?._el?.dataset.isManualLineBreak &&
+    getController(mq).cursor[dir]?.[dir]?._el?.dataset.isManualLineBreak &&
+    getController(mq).cursor[dir]?.[dir]?.[dir]?._el?.dataset.isManualLineBreak
```
sure lol

radian628 replied:
lmao

jared-hughes commented on src/plugins/multiline/index.ts:
```diff
@@ -136,6 +144,14 @@ export default class Multiline extends PluginController<Config> {
+          determineLineBreaksAutomatically:
+            this.settings.automaticallyMultilinify &&
+            (this.settings.disableAutomaticLineBreaksForHandAlignedExpressions
+              ? !(
+                  f?.parentElement?._mqMathFieldInstance?.latex?.() as string
+                ).match(/\\ \\ \\/g)
```
Instead of `str.match(/\\ \\ \\/g)`, I'd prefer `/\\ \\ \\/`.test(str) since that is intended to return a boolean.

radian628 replied:
Oh good point, I didn't know there was a function that did that lol

radian628 replied:
fixed

jared-hughes commented on src/plugins/multiline/index.ts:
```diff
@@ -160,6 +176,53 @@ export default class Multiline extends PluginController<Config> {
+          }
+
+          if (mq && key === "Backspace" && this.settings.spacesToNewlines) {
+            if (isNextToTripleSpaceLineBreak(mq, -1)) {
```
Remark: Mathquill lets `/** left */ const L = -1` and `/** right */ const R = +1`, so there's not just magical -1 and +1 everywhere.

radian628 replied:
Added.

jared-hughes commented on src/plugins/multiline/index.ts:
```diff
@@ -160,6 +176,53 @@ export default class Multiline extends PluginController<Config> {
+            }
+          }
+
+          // handle arrow nav with spaces2newlines
```
We'll have to see if this works well with Better Navigation.

radian628 replied:
Yep

jared-hughes commented on src/plugins/multiline/verticalify.ts:
```diff
@@ -109,10 +115,11 @@ export function unverticalify(elem: Element, force?: boolean) {
 
       // revert linebreaks to original symbol to get rid of <br>
-      if (child.dataset.isLineBreak) {
+      if (child.dataset.isLineBreak ?? child.dataset.isManualLineBreak) {
```
Might be worth renaming `isLineBreak` to `isAutoLineBreak` just to be clear about the distinction.

radian628 replied:
Agreed.

jared-hughes commented on src/plugins/multiline/verticalify.ts:
```diff
@@ -246,7 +304,7 @@ export function verticalify(
 
     // verticalify child
-    if (width > options.skipWidth) {
+    if (width > options.skipWidth || options.spacesToNewlines) {
```
Might need some in
```suggestion
    if (
      (options.determineLineBreaksAutomatically && width > options.skipWidth) ||
      options.spacesToNewlines
    ) {
```

jared-hughes commented on src/plugins/multiline/config.ts:
```diff
@@ -9,10 +9,27 @@ export const configList = [
     step: 1,
     key: "widthBeforeMultiline",
```
`widthBeforeMultiline` should only show if `determineLineBreaksAutomatically` is true, at least how I understand it (it only affects the cutoff for determining line breaks).

radian628 replied:
Yeah, your understanding is right here. `widthBeforeMultiline` is only relevant if `determineLineBreaksAutomatically` is true.

radian628 replied:
fixed

radian628 commented:
I've made a few more changes to the localizations as well, mostly just moving the stuff about the `Ctrl+M` hotkey to under the relevant settings in order to avoid any potential confusion between `Ctrl+M` and `Shift+Enter`.

jared-hughes commented:
Overflow issue here. Don't think it's avoidable though. Maybe treating the number field as inline, and using `&nbsp;` to avoid breaking between "Threshold" and "(%)". But not worthwhile.

![image](https://github.com/DesModder/DesModder/assets/20214911/75bc5e01-4a31-4ba1-9727-378e6ca8ddd0)

radian628 commented:
Settings menu is due for an overhaul anyway imo with the number of plugins we have now.

jared-hughes commented:
Agree about the need for a settings menu overhaul. Some ideas at https://github.com/DesModder/DesModder/issues/661

jared-hughes commented:
Just needs a main-merge and a check if this works with Ctrl+Arrow. Then it should be good to merge.

radian628 commented:
Looks like it breaks with Ctrl+Shift+Arrow rn. Probably an easy fix.

radian628 commented:
Fixed Ctrl+Shift+Arrow with multiline and also added Delete support. One weird wrinkle of this is that regardless of whether better nav is enabled, line breaks respond to keypresses with Ctrl. I don't think this is that big a deal since these key combinations don't do anything without better nav so they can't potentially interfere with anything.

jared-hughes commented:
What do you mean by "regardless of whether better nav is enabled, line breaks respond to keypresses with Ctrl"?

jared-hughes commented:
Bug:
1. Disable "Insert Linebreaks while Typing" or simply wait for the edit tdelay to finish after step 2
2. Type "( a b c Shift-Enter d e f )"
3. Press Ctrl+Z
The mathquill gets undo'd to a state with a triple-space that's not treated as a line break.

radian628 commented:
My bad that was poorly explained lol. The change I made is that navigating through line breaks (the Shift+Enter triple-space kind) is possible regardless of whether you're holding down Ctrl or not and regardless of whether better navigation is enabled. On the other hand, when navigating through anything else with betternav _disabled_, Ctrl prevents the arrow keys from doing anything. This makes it slightly weird because if better nav is disabled, something like, say, Ctrl+Backspace will delete a triple-space newline, but Ctrl+Backspace won't delete anything else. Overall, as far as I can tell, I think this is so niche that it's hardly worth caring about, but I felt it was worth mentioning either way in case there _is_ some unforseen issue it could cause.

radian628 commented:
Hmm yeah I should probably force multilining on undo/redo events.

radian628 commented:
Managed to fix the undo bug. It's a bit clunky and could probably be optimized somewhat, but spamming undo/redo refreshes graph state anyway so I don't consider it a major issue.
